Алемасцев Никита

nikovOne opened the pull request:
-Создал необходимые классы
-Написал основной алгоритм расстановки слов по спирали
-Приложил тесты
-Решил 2 и 3 задачи

nikovOne committed:
-Создал необходимые классы
-Написал основной алгоритм расстановки слов по спирали
-Приложил тесты
-Решил 2 и 3 задачи

nikovOne commented:
@shemloft

nikovOne committed:
-Добавил метод сдвига каждого нового прямоугольника
-Почистил код
-Чуть расширил тесты
-Добавил в вывод изображения ошибочного теста порядковые номера прямоугольников

nikovOne committed:
-Удалил ненужные using

shemloft commented on cs/TagsCloudVisualization/README.md:
```diff
@@ -0,0 +1,3 @@
+﻿В Program.cs есть список слов. Можно загрузить любое количество слов для получения нового облака. Также перезапуск с тем же набором слов
+приведёт к другому результату, т.к. в алгоритме есть случайная величина для размера шрифта каждого слова.
+![](bin/Debug/Sample.png)
```
Заработает, только если локально открыть ридми, предварительно запустив программу. Сохрани потом примеры картинок так, чтобы можно было из гита посмотреть

shemloft replied:
Всё еще не сделано,  в файле ридми нет изображений, в ТЗ есть четкое требование:
> сохрани изображения в репозитории и создай в директории проекта README.md файл включающий эти изображения

shemloft commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,41 @@
+            
+            var random = new Random();
+            var image = new Bitmap(6000, 3000);
+            var circ = new CircularCloudLayouter(new Point(image.Width / 2, image.Height / 2));
```
При кратком названии переменной обычно используют основное существительное, которое описывает класс. В данном случае это layouter. Circ это начало прилагательного, характеризующего этот конкретный вид layouter-а

shemloft commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,41 @@
+            var brush = Graphics.FromImage(image);
+            var pen = new Pen(Color.Black);
+            
+            brush.DrawEllipse(new Pen(Color.Red, 3), image.Width / 2, image.Height / 2, 3, 3);
```
Второй раз используешь точку, определяющую центр, лучше вынести в переменную

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+{
+    internal class CircularCloudLayouter
+    {
+        public Rectangle filledArea;
```
Использовать публичные поля плохая практика. Для всего публичного лучше использовать свойства, или делать такие поля readonly. Ну и все публичное должно быть с большой буквы, решарпер это подсказывает

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+    internal class CircularCloudLayouter
+    {
+        public Rectangle filledArea;
+        public List<Rectangle> Rectangles = new List<Rectangle>();
```
Это поле не должно торчать наружу. Во-первых, в него можно добавить новые прямоугольники вне класса, что не очень хорошо. Во-вторых оно используется только в тестах. Бизнес код не должен подстраиваться под тесты. Аналогично для предыдущего поля

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+            minLeft = center.X;
+        }
+
+        public Rectangle PullNextRectangle(Size rectangleSize)
```
Pull это вытянуть.. В ТЗ было Put, т.е. поставить

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+        private int maxBottom;
+        private int minTop;
+        private int minLeft;
+        private int maxRight;
```
Названия странные. Если есть maxBottom, то где minBottom? Ты описываешь точку двумя словами, которые сами по себе каждое могло бы ее идентифицировать. Или maxY-minY, или filledAreaBottom-filledAreaTop. Выбери что-то одно.

Да и вообще, зачем хранить отдельно координаты прямоугольника и сам прямоугольник. Храни их сразу в прямоугольнике, зачем плодить лишние переменные

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+    {
+        public Rectangle filledArea;
+        public List<Rectangle> Rectangles = new List<Rectangle>();
+        private Rectangle? previousRectangle = null;
```
Вместо этого можно вполне использовать `Rectangles.LastOrDefault()`.

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+        private void UpdateSizeFilledRect(Rectangle nextRect)
+        {
+            if (nextRect.Bottom > maxBottom)
+                maxBottom = nextRect.Bottom;
```
Вкусовщина, но мне нравится в таких случаях использовать Math.Max и Math.Min. С их использованием здесь будет 
`maxBottom = Math.Max(maxBottom, next.RectBottom)`

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+
+        private void CheckAndSwitchSideToFill(Size rectangleSize, Rectangle nextRect, Point nextPoint)
+        {
+            if (sideToFill == SideToFill.Up && (nextRect.Right - filledArea.Right) / (double)nextRect.Width >= 0.5)
```
Лучше такие вычисления выносить в именованную переменную. Код должен читаться как текст, а чтобы разобраться в том, что значат эти формулы, нужно посидеть и подумать

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+
+            var newPoint = new Point();
+            
+            if (sideToFill is SideToFill.Up)
```
С enum-ами лучше использовать switch

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+            return newPoint;
+        }
+
+        private Rectangle Move(Rectangle nextRectangle)
```
Слишком длинный метод с неговорящим названием и неочевидными вычислениями. Лучше разбить на несколько

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+            if (maxPossibleDistanceRight > 0 && maxPossibleDistanceRight != int.MaxValue && sideToFill is SideToFill.Left)
+                nextRectangle.X += maxPossibleDistanceRight;
+            if (maxPossibleDistanceUp > 0 && maxPossibleDistanceUp != int.MaxValue && sideToFill is SideToFill.Down)
+                nextRectangle.Y -= maxPossibleDistanceUp;
```
Ты в итоге используешь только одно значение дельты в зависимости от SideToFill, но считаешь все 4. Считай только нужную, остальные вычисления же никак не сохраняются

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+            return nextRectangle;
+        }
+
+        private bool HaveIntersectionOnX(Rectangle rect1, Rectangle rect2)
```
Неочевидно где строгое неравенство, где нестрогое и вообще непонятно, что тут происходит. Если ты хочешь проверить не пересекаются ли отрезки (a, b) и (c, d), то есть более простой метод: пересекаются, если `max(a,c)<=min(b,d)`

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,195 @@
+            if (sideToFill == SideToFill.Up && (nextRect.Right - filledArea.Right) / (double)nextRect.Width >= 0.5)
+            {
+                sideToFill = SideToFill.Right;
+                UpdateFilledRect();
```
Этот метод вызывается в каждом if-е, его вполне можно вынести за пределы if-а

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,147 @@
+        }
+        
+        [TearDown]
+        public void E()
```
![E](https://user-images.githubusercontent.com/38860225/140804435-8283e685-c846-46d0-9df2-36bab0dfc169.jpg)
так методы не называют

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,147 @@
+    {
+        private CircularCloudLayouter cloudLayouter;
+
+        private static List<List<int>> rectanglesSampleForPlaceCheck = new List<List<int>>
```
Используется ровно один раз, лучше создавать прямо в тесте. И сделать списки из используемых классов, а не из интов. Аналогично для списка ниже

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,147 @@
+        [Test]
+        public void PullNextRectangle_ShouldPlace()
+        {
+            foreach (var source in rectanglesSampleForPlaceCheck)
```
Нарушается AAA

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        [Test]
+        public void PullNextRectangle_ShouldPlace()
```
Обычно в подобном именовании тестов есть три части: Метод_ЧтоДолженСделать_Условие.
Здесь не хватает конкретики, что именно проверяет тест

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        [Test]
+        public void PullNextRectangle_ShouldPlaceFirstRectangleLeftFromCenter()
```
Обычно в подобном именовании тестов есть три части: Метод_ЧтоДолженСделать_Условие.
Тут все есть, но не разбито по частям

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,147 @@
+            var size = new Size(30, 40);
+            var cloudLayouter = new CircularCloudLayouter(center);
+            var rect = new Rectangle(new Point(-size.Width, 0), size);
+            cloudLayouter.PullNextRectangle(size).Should().BeEquivalentTo(rect);
```
Части AAA обычно отделяются друг от друга пустой строкой. Здесь совмещены act и assert, это норм, но отделить arrange стоит. В других местах тоже

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,147 @@
+        }
+        
+        [Test]
+        public void UpdateFilledRect_ShouldBe()
```
Те же комменты, что и для теста PullNextRectangle_ShouldPlace

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,147 @@
+            {
+                for (var j = i + 1; j < cloudLayouter.Rectangles.Count; j++)
+                {
+                    Assert.False(cloudLayouter.Rectangles[i].IntersectsWith(cloudLayouter.Rectangles[j]));
```
Если начал использовать FLuentAssertions, используй до конца)

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,147 @@
+        [TestCase(10, 0, TestName = "Zero height")]
+        public void PullNextRectangle_ShouldThrowArgumentException_WhenSizeIncorrect(int width, int height)
+        {
+            Action act = () => cloudLayouter.PullNextRectangle(new Size(-10, 10));
```
Кажется, ты забыл заиспользовать тест кейсы 🤔

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        [Test]
+        public void Test_ShouldFail_ForPictureSave()
```
Вижу, что работает сохранение картинки, но лучше не оставлять падающих тестов

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,147 @@
+namespace TagsCloudVisualization
+{
+    [TestFixture]
+    public class CircularCloudLayouterTests
```
Стоит добавить больше тестов на пересечение, желательно с рандомным входом, а так же тест на плотность облака

shemloft commented:
Стоит сделать облако более круглым, оно сильно разрастается в стороны

shemloft commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,41 @@
+            brush.DrawEllipse(new Pen(Color.Red, 3), image.Width / 2, image.Height / 2, 3, 3);
+            foreach (var word in words)
+            {
+                var fontSize = random.Next(16, 48);
```
Стоит инкапсулировать логику отрисовки

shemloft replied:
Логика отрисовки разрослась, появилась куча дублирований. В этом комментарии имелось ввиду, что необходимо вынести отрисовку в отдельный класс

shemloft reviewed (commented):
Стоит сделать облако более круглым, оно сильно разрастается в стороны. Так же, кажется если добавить сначала маленький прямоугольник, потом большой, потом много маленьких, решение сломается

nikovOne committed:
-Добавил класс Movement
-Добавил enum MovementType и Quarter
-Переработал добавление прямоугольников, сделал с помощью спирали Архимеда
-Добавил тест на плотность
-Увеличил тест на пересечения прямоугольников
-Доработал класс Program.cs
-Прочие мелкие доработки

nikovOne committed:
Добавил примеры работы

shemloft commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,69 @@
+{
+    internal class Program
+    {
+        public static void Main(string[] args)
```
Метод огромный. Если метод получается больше 15 строк, то его скорее всего стоит разбить на подметоды.

shemloft commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,69 @@
+            Console.WriteLine("1 - облако слов из листа");
+            Console.WriteLine("2 - случайный размер от 5 до 70 по каждой оси");
+            Console.WriteLine("3 - фиксированный размер 40");
+            var answer = int.Parse(Console.ReadLine() ?? throw new InvalidOperationException());
```
Странный интерактив, достаточно просто было нагенерировать эти случаи и сохранить изображения. Ну и всего три случая, везде забиты константы. 
В целом по коду смешана генерация тестовых данных и отрисовка
Отрисовку вообще нужно вынести в отдельный класс

shemloft commented on cs/TagsCloudVisualization/MovementType.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace TagsCloudVisualization
+{
+    public enum MovementType
```
MovementDirection или просто Direction

shemloft commented on cs/TagsCloudVisualization/Movement.cs:
```diff
@@ -0,0 +1,45 @@
+            this.distance = distance;
+        }
+
+        public bool CanDoMovement()
```
CanMove

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,173 @@
+        public CircularCloudLayouter(Point center)
+        {
+            this.center = center;
+            step = 10;
```
1. step и angle используются ровно в одном методе - GetNextPoint. Выглядит как математика, которую можно вынести в отдельный класс, т.к. кроме этих двух константы и переменной там из класса используется только center как сдвиг, его можно передавать как аргумент или уже в самом Layouter применять
2. step это константа, ее нет смысла задавать в конструкторе и навешивать readonly
3. angle-у тоже лучше задать начальное значение на этапе объявления

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,173 @@
+            angle = 0;
+        }
+
+        public Rectangle[] GetPutRectangles()
```
Какой дополнительный смысл несет Put по середине? Почему не просто GetRectangles?

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,173 @@
+    internal class CircularCloudLayouter
+    {
+        private readonly List<Rectangle> rectangles = new List<Rectangle>();
+        private Point center;
```
Поле не меняется, сделай или public readonly или get-only свойством и не добавляй лишний метод GetCenter

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,173 @@
+            return center;
+        }
+
+        public Rectangle PutNextRectangle(Size rectangleSize)
```
Метод слишком большой, и тут внутри есть явных логических части, которые так и просятся, чтобы их выделили в отдельные методы

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,173 @@
+                    $"Size is empty! Width: {rectangleSize.Width}, height: {rectangleSize.Height}");
+            }
+            
+            Rectangle nextRectangle;
```
Это и do..while. Мб и позволяет не писать чтение точки вне цикла, но выглядит страшно и сложно читается. Лучше не использовать do..while, если можно использовать что-то другое.

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,173 @@
+                var previousQuarter = GetQuarter(nextRectangle);
+                nextRectangle = possibleMovementToCenter.MoveRectangle(nextRectangle);
+                
+                if (QuarterChanged(previousQuarter, nextRectangle))
```
А вот проверку в одну строчку выносить в метод не обязательно. Читаемость никуда не денется, если просто написать условие

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,173 @@
+            return new Point(x, y);
+        }
+
+        private bool DoesIntersectPreviousRectangles(Rectangle rectangle)
```
DoesIntersect => Intersects

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,173 @@
+
+        private static bool CanMadeMoveOnDistance(int distance)
+        {
+            return distance > 0 && distance != int.MaxValue;
```
Точно такой же метод есть в классе Movement. Что-то тут не так

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,173 @@
+            return distance > 0 && distance != int.MaxValue;
+        }
+        
+        private Movement GetPossibleMovement(Rectangle rectangle)
```
Слишком большой => можно выделить подметоды

shemloft replied:
И все ветки очень похожи, постарайся убрать дублирование

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,122 @@
+        }
+
+        [Test]
+        public void PullNextRectangle_ShouldReturn_NotIntersectedRects()
```
Название не поменялось, а проверяемый метод поменялся

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,122 @@
+            var random = new Random();
+            
+            for (var i = 0; i < 100; i++)
+                cloudLayouter.PutNextRectangle(new Size(random.Next(5, 70), random.Next(5, 70)));
```
Дублирование кода здесь и в другом тесте

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,173 @@
+            return result;
+        }
+
+        private Quarter GetQuarter(Rectangle rectangle)
```
Зачем передавать весь прямоугольник, когда используется только его Location

shemloft replied:
Выглядит как метод, который можно вынести в какой-нибудь Extension или Helper, с сутью вроде "отношение точки к началу координат" куда передается точка и начало координат. И вообще в целом стоит повыносить методы, которые более общие, чем потрености layouter-а

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,122 @@
+
+            var rectangles = cloudLayouter.GetPutRectangles();
+
+            for (var i = 0; i < rectangles.Length - 1; i++)
```
Вынеси в отдельный метод сбор пересечений

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,122 @@
+                cloudLayouter.PutNextRectangle(new Size(random.Next(5, 70), random.Next(5, 70)));
+
+            var rectangles = cloudLayouter.GetPutRectangles();
+
```
А что за четвертая часть отделена пустой строкой? В AAA три части

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,122 @@
+                cloudLayouter.PutNextRectangle(new Size(random.Next(5, 70), random.Next(5, 70)));
+            
+            var rectangles = cloudLayouter.GetPutRectangles();
+            var averageDistanceFromCenterToBorder = GetAverageDistanceFromCenterToBorder(center, rectangles);
```
Вынеси расчеты в отдельный метод с понятным названием

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,122 @@
+namespace TagsCloudVisualization
+{
+    [TestFixture]
+    public class CircularCloudLayouterTests
```
Нет тестов на форму облака:
> Форма итогового облака должна быть близка к кругу с центром в точке center
> Все требования должны вычитываться из твоих модульных тестов

shemloft commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,122 @@
+            }
+        }
+
+        private static void DrawRectangles(IReadOnlyList<Rectangle> rectangles, Graphics brush)
```
Приватные методы лучше размещать после всех публичных

shemloft commented:
Тесты стоит выделять в отдельный .csproj, и все проекты в отдельный .sln
